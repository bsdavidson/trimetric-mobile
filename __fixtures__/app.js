export const props = {
  dimensions: {
    screen: {
      width: 500,
      height: 600
    }
  },
  following: false,
  layerVisibility: {},
  loaded: true,
  routeShapes: [],
  selectedArrival: {},
  selectedArrivalVehicleInfo: {},
  selectedItem: {},
  selectedItemsInfo: [],
  selectedItemsViewHeight: 100,
  selectedVehicleInfo: {},
  stopPoints: [],
  vehiclePoints: []
};
