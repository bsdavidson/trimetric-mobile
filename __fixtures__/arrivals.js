export const selectedStop = {
  type: "stop",
  position: [-122.68931172788143, 45.51957800931885],
  item: {
    type: "Feature",
    properties: {stop_id: "5016", type: "stop"},
    geometry: {
      type: "Point",
      coordinates: [-122.68931172788143, 45.51957800931885]
    }
  }
};

export const selectedItem = {
  type: "stop",
  position: [-122.68931172788143, 45.51957800931885],
  item: {
    type: "Feature",
    properties: {stop_id: "5016", type: "stop"},
    geometry: {
      type: "Point",
      coordinates: [-122.68931172788143, 45.51957800931885]
    }
  }
};

export const arrivals = [
  {
    route_id: "51",
    route_short_name: "51",
    route_long_name: "Vista",
    route_type: 3,
    route_color: "",
    route_text_color: "",
    trip_id: "7908833",
    stop_id: "5016",
    headsign: "Portland",
    arrival_time: "16:22:00",
    departure_time: "16:22:00",
    vehicle_id: null,
    vehicle_label: null,
    vehicle_position: {
      lat: 0,
      lng: 0,
      bearing: 0,
      odometer: 0,
      speed: 0
    },
    date: "2018-01-19T00:00:00Z",
    trip_shape: {
      direction_id: 1,
      route_id: "51",
      color: "",
      points: [
        {lat: 45.493347, lng: -122.70053},
        {lat: 45.493347, lng: -122.700701},
        {lat: 45.516844, lng: -122.678826}
      ],
      trip_id: "7908833"
    }
  },
  {
    route_id: "51",
    route_short_name: "51",
    route_long_name: "Vista",
    route_type: 3,
    route_color: "",
    route_text_color: "",
    trip_id: "7908834",
    stop_id: "5016",
    headsign: "Portland",
    arrival_time: "16:52:00",
    departure_time: "16:52:00",
    vehicle_id: null,
    vehicle_label: null,
    vehicle_position: {
      lat: 0,
      lng: 0,
      bearing: 0,
      odometer: 0,
      speed: 0
    },
    date: "2018-01-19T00:00:00Z",
    trip_shape: {
      direction_id: 1,
      route_id: "51",
      color: "",
      points: [
        {lat: 45.490581, lng: -122.709817},
        {lat: 45.490589, lng: -122.710375},
        {lat: 45.516844, lng: -122.678826}
      ],
      trip_id: "7908834"
    }
  }
];

export const selectedItems = {
  type: "FeatureCollection",
  features: [
    {
      type: "Feature",
      properties: {stop_id: "5016", type: "stop"},
      geometry: {
        type: "Point",
        coordinates: [-122.68931172788143, 45.51957800931885]
      }
    }
  ]
};
