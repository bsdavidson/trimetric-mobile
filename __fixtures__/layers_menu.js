export const display = true;

export const bottom = 60;

export const selectedItemsViewHeight = 0;

export const layerVisibility = {
  routeShapes: true,
  buses: false,
  trains: true,
  stops: true,
  vehicleLabels: true,
  buildings: false
};

export const dimensions = {
  window: {fontScale: 1, width: 375, height: 667, scale: 2},
  screen: {fontScale: 1, width: 375, height: 667, scale: 2}
};
