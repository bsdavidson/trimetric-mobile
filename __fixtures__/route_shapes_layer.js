export const props = {
  visible: true,
  routeShapes: {
    type: "FeatureCollection",
    features: [
      {
        type: "Feature",
        properties: {color: "#CE0F69", route_id: "194"},
        geometry: {
          type: "MultiLineString",
          coordinates: [
            [
              [-122.680501, 45.53012],
              [-122.681354, 45.530079],
              [-122.681105, 45.530084]
            ]
          ]
        }
      },
      {
        type: "Feature",
        properties: {color: "#84BD00", route_id: "193"},
        geometry: {
          type: "MultiLineString",
          coordinates: [
            [
              [-122.671375, 45.493955],
              [-122.671107, 45.49395],
              [-122.698618, 45.529861],
              [-122.698646, 45.530509],
              [-122.698649, 45.530616]
            ],
            [
              [-122.685036, 45.514564],
              [-122.685257, 45.514157],
              [-122.685323, 45.514034],
              [-122.685339, 45.514005]
            ]
          ]
        }
      }
    ]
  },
  selectedArrival: null
};
