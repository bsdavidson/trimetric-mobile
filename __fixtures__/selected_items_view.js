// Selected Stop
export const stopData = [
  {
    type: "stop",
    stop: {
      id: "7749",
      code: "7749",
      name: "SW 6th & Broadway",
      desc: "Northbound stop in Portland (Stop ID 7749)",
      lat: 45.506859,
      lng: -122.684151,
      zone_id: "B",
      url: "http://trimet.org/#tracker/stop/7749",
      location_type: 0,
      parent_station: "",
      direction: "North",
      position: "Nearside",
      wheelchair_boarding: 0,
      distance: 0
    },
    arrivals: [
      {
        route_id: "8",
        route_short_name: "8",
        route_long_name: "Jackson Park/NE 15th",
        route_type: 3,
        route_color: "",
        route_text_color: "",
        trip_id: "7902197",
        stop_id: "7749",
        headsign: "NE Dekum",
        arrival_time: "14:12:07",
        departure_time: "14:12:07",
        vehicle_id: "2652",
        vehicle_label: "8  To NE Dekum",
        vehicle_position: {
          lat: 45.498703,
          lng: -122.68535,
          bearing: 224,
          odometer: 0,
          speed: 0
        },
        date: "2018-01-22T00:00:00Z",
        trip_shape: {
          direction_id: 1,
          route_id: "8",
          color: "",
          points: [
            {lat: 45.499398, lng: -122.685589},
            {lat: 45.571787, lng: -122.660853}
          ],
          trip_id: "7902197"
        },
        nextArrivals: [
          {
            route_id: "8",
            route_short_name: "8",
            route_long_name: "Jackson Park/NE 15th",
            route_type: 3,
            route_color: "",
            route_text_color: "",
            trip_id: "7902198",
            stop_id: "7749",
            headsign: "NE Dekum",
            arrival_time: "14:22:07",
            departure_time: "14:22:07",
            vehicle_id: "3142",
            vehicle_label: "8  To NE Dekum",
            vehicle_position: {
              lat: 45.497105,
              lng: -122.68317,
              bearing: 208,
              odometer: 0,
              speed: 0
            },
            date: "2018-01-22T00:00:00Z",
            trip_shape: {
              direction_id: 1,
              route_id: "8",
              color: "",
              points: [
                {lat: 45.496432, lng: -122.682258},
                {lat: 45.571787, lng: -122.660853}
              ],
              trip_id: "7902198"
            }
          },
          {
            route_id: "8",
            route_short_name: "8",
            route_long_name: "Jackson Park/NE 15th",
            route_type: 3,
            route_color: "",
            route_text_color: "",
            trip_id: "7902199",
            stop_id: "7749",
            headsign: "NE Dekum",
            arrival_time: "14:32:07",
            departure_time: "14:32:07",
            vehicle_id: null,
            vehicle_label: null,
            vehicle_position: {
              lat: 0,
              lng: 0,
              bearing: 0,
              odometer: 0,
              speed: 0
            },
            date: "2018-01-22T00:00:00Z",
            trip_shape: {
              direction_id: 1,
              route_id: "8",
              color: "",
              points: [
                {lat: 45.499398, lng: -122.685589},
                {lat: 45.571787, lng: -122.660853}
              ],
              trip_id: "7902199"
            }
          },
          {
            route_id: "8",
            route_short_name: "8",
            route_long_name: "Jackson Park/NE 15th",
            route_type: 3,
            route_color: "",
            route_text_color: "",
            trip_id: "7902200",
            stop_id: "7749",
            headsign: "NE Dekum",
            arrival_time: "14:42:44",
            departure_time: "14:42:44",
            vehicle_id: null,
            vehicle_label: null,
            vehicle_position: {
              lat: 0,
              lng: 0,
              bearing: 0,
              odometer: 0,
              speed: 0
            },
            date: "2018-01-22T00:00:00Z",
            trip_shape: {
              direction_id: 1,
              route_id: "8",
              color: "",
              points: [
                {lat: 45.499398, lng: -122.685589},
                {lat: 45.571787, lng: -122.660853}
              ],
              trip_id: "7902200"
            }
          }
        ]
      },
      {
        route_id: "8",
        route_short_name: "8",
        route_long_name: "Jackson Park/NE 15th",
        route_type: 3,
        route_color: "",
        route_text_color: "",
        trip_id: "7902198",
        stop_id: "7749",
        headsign: "NE Dekum",
        arrival_time: "14:22:07",
        departure_time: "14:22:07",
        vehicle_id: "3142",
        vehicle_label: "8  To NE Dekum",
        vehicle_position: {
          lat: 45.497105,
          lng: -122.68317,
          bearing: 208,
          odometer: 0,
          speed: 0
        },
        date: "2018-01-22T00:00:00Z",
        trip_shape: {
          direction_id: 1,
          route_id: "8",
          color: "",
          points: [
            {lat: 45.496432, lng: -122.682258},
            {lat: 45.571787, lng: -122.660853}
          ],
          trip_id: "7902198"
        }
      },
      {
        route_id: "8",
        route_short_name: "8",
        route_long_name: "Jackson Park/NE 15th",
        route_type: 3,
        route_color: "",
        route_text_color: "",
        trip_id: "7902199",
        stop_id: "7749",
        headsign: "NE Dekum",
        arrival_time: "14:32:07",
        departure_time: "14:32:07",
        vehicle_id: null,
        vehicle_label: null,
        vehicle_position: {
          lat: 0,
          lng: 0,
          bearing: 0,
          odometer: 0,
          speed: 0
        },
        date: "2018-01-22T00:00:00Z",
        trip_shape: {
          direction_id: 1,
          route_id: "8",
          color: "",
          points: [
            {lat: 45.499398, lng: -122.685589},
            {lat: 45.571787, lng: -122.660853}
          ],
          trip_id: "7902199"
        }
      },
      {
        route_id: "8",
        route_short_name: "8",
        route_long_name: "Jackson Park/NE 15th",
        route_type: 3,
        route_color: "",
        route_text_color: "",
        trip_id: "7902200",
        stop_id: "7749",
        headsign: "NE Dekum",
        arrival_time: "14:42:44",
        departure_time: "14:42:44",
        vehicle_id: null,
        vehicle_label: null,
        vehicle_position: {
          lat: 0,
          lng: 0,
          bearing: 0,
          odometer: 0,
          speed: 0
        },
        date: "2018-01-22T00:00:00Z",
        trip_shape: {
          direction_id: 1,
          route_id: "8",
          color: "",
          points: [
            {lat: 45.499398, lng: -122.685589},
            {lat: 45.571787, lng: -122.660853}
          ],
          trip_id: "7902200"
        }
      },
      {
        route_id: "8",
        route_short_name: "8",
        route_long_name: "Jackson Park/NE 15th",
        route_type: 3,
        route_color: "",
        route_text_color: "",
        trip_id: "7902201",
        stop_id: "7749",
        headsign: "NE Dekum",
        arrival_time: "14:52:44",
        departure_time: "14:52:44",
        vehicle_id: null,
        vehicle_label: null,
        vehicle_position: {
          lat: 0,
          lng: 0,
          bearing: 0,
          odometer: 0,
          speed: 0
        },
        date: "2018-01-22T00:00:00Z",
        trip_shape: {
          direction_id: 1,
          route_id: "8",
          color: "",
          points: [
            {lat: 45.496432, lng: -122.682258},
            {lat: 45.571787, lng: -122.660853}
          ],
          trip_id: "7902201"
        }
      },
      {
        route_id: "8",
        route_short_name: "8",
        route_long_name: "Jackson Park/NE 15th",
        route_type: 3,
        route_color: "",
        route_text_color: "",
        trip_id: "7902202",
        stop_id: "7749",
        headsign: "NE Dekum",
        arrival_time: "15:02:44",
        departure_time: "15:02:44",
        vehicle_id: null,
        vehicle_label: null,
        vehicle_position: {
          lat: 0,
          lng: 0,
          bearing: 0,
          odometer: 0,
          speed: 0
        },
        date: "2018-01-22T00:00:00Z",
        trip_shape: {
          direction_id: 1,
          route_id: "8",
          color: "",
          points: [
            {lat: 45.499398, lng: -122.685589},
            {lat: 45.571787, lng: -122.660853}
          ],
          trip_id: "7902202"
        }
      }
    ]
  }
];

export const following = true;
export const selectedItem = {
  type: "stop",
  position: [-122.68415115773678, 45.50685913695389],
  item: {
    type: "Feature",
    properties: {stop_id: "7749", type: "stop"},
    geometry: {
      type: "Point",
      coordinates: [-122.68415115773678, 45.50685913695389]
    }
  }
};

export const selectedItemIndex = 0;
export const selectedArrival = null;

export const selectedArrivalVehicleInfo = null;
export const selectedStopInfo = {
  id: "7749",
  code: "7749",
  name: "SW 6th & Broadway",
  desc: "Northbound stop in Portland (Stop ID 7749)",
  lat: 45.506859,
  lng: -122.684151,
  zone_id: "B",
  url: "http://trimet.org/#tracker/stop/7749",
  location_type: 0,
  parent_station: "",
  direction: "North",
  position: "Nearside",
  wheelchair_boarding: 0,
  distance: 0
};

// Selected Vehicle
export const vehicleData = [
  {
    type: "vehicle",
    vehicle: {
      trip: {trip_id: "7918798", route_id: "200"},
      vehicle: {id: "206", label: "Green Line to Clackamas"},
      position: {
        lat: 45.50877,
        lng: -122.68381,
        bearing: 331,
        odometer: 0,
        speed: 0
      },
      current_stop_sequence: 1,
      stop_id: "13140",
      current_status: 1,
      timestamp: 1516660571,
      congestion_level: 0,
      occupancy_status: 0,
      route_type: 0
    }
  }
];

export const selectedItemVehicle = {
  type: "vehicle",
  position: [-122.68381, 45.50877],
  item: {
    type: "Feature",
    properties: {
      bearing: 331,
      route_id: "200",
      icon: "tram",
      vehicle_id: "206",
      type: "vehicle",
      stop_id: "13140"
    },
    geometry: {type: "Point", coordinates: [-122.68381, 45.50877]}
  }
};

export const selectedVehicleInfo = {
  id: "13140",
  code: "13140",
  name: "Jackson St Turnaround",
  desc: "Westbound stop in Portland (Stop ID 13140)",
  lat: 45.508653,
  lng: -122.6836,
  zone_id: "R",
  url: "http://trimet.org/#tracker/stop/13140",
  location_type: 0,
  parent_station: "",
  direction: "West",
  position: "At",
  wheelchair_boarding: 0,
  distance: 0
};
