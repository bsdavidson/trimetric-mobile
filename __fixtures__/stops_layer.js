export const stopPoints = {
  type: "FeatureCollection",
  features: [
    {
      type: "Feature",
      properties: {type: "stop", stop_id: "13990"},
      geometry: {type: "Point", coordinates: [-122.398702, 45.552013]}
    },
    {
      type: "Feature",
      properties: {type: "stop", stop_id: "13991"},
      geometry: {type: "Point", coordinates: [-122.407683, 45.55363]}
    }
  ]
};

export const stops = [
  {
    id: "2",
    code: "2",
    name: "A Ave & Chandler",
    desc: "Eastbound stop in Lake Oswego (Stop ID 2)",
    lat: 45.420609,
    lng: -122.675671,
    zone_id: "B",
    url: "http://trimet.org/#tracker/stop/2",
    location_type: 0,
    parent_station: "",
    direction: "East",
    position: "Nearside",
    wheelchair_boarding: 0,
    distance: 0
  },
  {
    id: "landmark-2560",
    code: "",
    name: "SE Park Ave MAX Station",
    desc: "",
    lat: 45.430735,
    lng: -122.635065,
    zone_id: "",
    url: "",
    location_type: 1,
    parent_station: "",
    direction: "",
    position: "",
    wheelchair_boarding: 0,
    distance: 0
  }
];
