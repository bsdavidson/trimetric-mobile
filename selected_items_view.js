import React, {Component} from "react";
import {connect} from "react-redux";
import Moment from "moment";

import {getSelectedItem, ROUTE_TYPE_ICONS} from "./selectors";

import {
  Animated,
  Dimensions,
  FlatList,
  Image,
  PixelRatio,
  Platform,
  StyleSheet,
  Text,
  TouchableHighlight,
  TouchableOpacity,
  View
} from "react-native";

import {selectItemIndex, setMapViewInset} from "./actions";
import Arrivals from "./arrivals";
import tram from "./assets/tram.png";
import bus from "./assets/bus.png";
import stopImage from "./assets/stop.png";

const VEHICLE_IMAGE = {
  0: tram,
  3: bus
};

const MAP_AREA_OFFSET = 250;
const HEADER_HEIGHT = 100;
const HEADER_PAGINATION_HEIGHT = 20;
const STATUS_BAR_OFFSET = Platform.OS === "android" ? 24 : 0;

function headerPagination(contentLength, index) {
  let pageNav = [];
  for (let i = 0; i < contentLength; i++) {
    let style = index === i ? styles.active : styles.inactive;
    pageNav.push(
      <View key={"pgnav-" + i} style={style}>
        <Text>{i + 1}</Text>
      </View>
    );
  }
  return pageNav;
}

function keyExtractor(item, index) {
  return String(index);
}

class SelectedItemsView extends Component {
  constructor(props) {
    super(props);
    let screenWidth = Dimensions.get("window").width;
    let screenHeight = Dimensions.get("window").height + STATUS_BAR_OFFSET;

    this.state = {
      currentIndex: 0,
      isOpen: false,
      layoutHeight: null,
      layoutWidth: null,
      marginTop: screenHeight * 0.3,
      openTween: new Animated.Value(0),
      screenHeight,
      screenWidth
    };

    this.headerListRef = null;

    this.calculatePageIndex = this.calculatePageIndex.bind(this);
    this.handleHeaderListRef = this.handleHeaderListRef.bind(this);
    this.handleHeaderScroll = this.handleHeaderScroll.bind(this);
    this.handleIconTap = this.handleIconTap.bind(this);
    this.handleLayout = this.handleLayout.bind(this);
    this.renderHeaderItem = this.renderHeaderItem.bind(this);
  }

  calculatePageIndex(itemWidth, currentOffset) {
    return Math.round(currentOffset / itemWidth);
  }

  componentDidMount() {
    this.close();
  }

  // compare the current selected item with the selected item(s) in the store
  // and return the index.
  getSelectedItemIndex() {
    const selectedItems = this.props.data;
    if (selectedItems.length === 0) {
      return -1;
    }
    const selectedItem = this.props.selectedItem;
    if (!selectedItem || !selectedItem[selectedItem.type]) {
      return -1;
    }

    for (let i = 0; i < selectedItems.length; i++) {
      const si = selectedItems[i];
      if (si.type !== selectedItem.type || !si[si.type]) {
        continue;
      }
      switch (selectedItem.type) {
        case "stop":
          if (si.stop.id === selectedItem.stop.id) {
            return i;
          }
          break;
        case "vehicle":
          if (si.vehicle.vehicle.id === selectedItem.vehicle.vehicle.id) {
            return i;
          }
          break;
        default:
          break;
      }
    }
    return -1;
  }

  // Handle what happens when you tap the image icon to raise and lower details
  handleIconTap() {
    this.openOrClose();
  }

  handleLayout(e) {
    let layoutWidth = e.nativeEvent.layout.width;
    let layoutHeight = e.nativeEvent.layout.height;
    let screenWidth = Dimensions.get("window").width;
    let screenHeight = Dimensions.get("window").height + STATUS_BAR_OFFSET;
    this.setState({
      layoutWidth,
      layoutHeight,
      screenWidth,
      screenHeight,
      marginTop: screenHeight * 0.3
    });

    // Fixes a bug that causes the drawer nav to sit on boundaries when the
    // layout is rotated.
    if (this.headerListRef) {
      this.headerListRef.scrollToIndex({index: 0});
    }
  }

  handleHeaderScroll(e) {
    let idx = this.calculatePageIndex(
      e.nativeEvent.layoutMeasurement.width,
      e.nativeEvent.contentOffset.x
    );
    if (idx !== this.props.selectedItemIndex) {
      this.selectItem(idx);
    }
  }

  handleHeaderListRef(list) {
    this.headerListRef = list;
  }

  openOrClose() {
    if (!this.state.isOpen) {
      this.open();
    } else {
      this.close();
    }
  }

  open() {
    // Raise drawer

    Animated.timing(this.state.openTween, {
      toValue: 1,
      duration: 300,
      useNativeDriver: true
    }).start();
    this.setState({
      isOpen: true
    });

    let height = this.state.screenHeight - this.state.marginTop;
    if (Platform.OS === "android") {
      // Android uses a different pixel measurement than ios
      height = Math.floor(PixelRatio.getPixelSizeForLayoutSize(height));
    }
    this.props.onResize(height);
  }

  close() {
    // close drawer

    Animated.timing(this.state.openTween, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true
    }).start();

    this.setState({
      isOpen: false
    });

    // Trigger layout resize
    this.props.onResize(0);
  }

  renderHeaderItem({item}) {
    if (item.type === "stop") {
      return (
        <StopItem
          width={this.state.layoutWidth}
          onPress={this.handleIconTap}
          stop={item.stop}
        />
      );
    }

    if (item.type === "vehicle") {
      return (
        <VehicleItem
          width={this.state.layoutWidth}
          onPress={this.handleIconTap}
          vehicle={item.vehicle}
        />
      );
    }
  }

  render() {
    let height = this.state.screenHeight - this.state.marginTop;
    let headerHeight = HEADER_HEIGHT;
    let header = null;
    if (this.props.data.length > 1) {
      headerHeight += HEADER_PAGINATION_HEIGHT;
      header = (
        <View style={styles.headerPagination}>
          {headerPagination(
            this.props.data.length,
            this.props.selectedItemIndex
          )}
        </View>
      );
    }
    let offset = Platform.OS === "android" ? 48 : 0;
    let viewStyle = {
      height: height,
      bottom: headerHeight - height,
      transform: [
        {
          translateY: this.state.openTween.interpolate({
            inputRange: [0, 1],
            outputRange: [
              0,
              -this.state.screenHeight + headerHeight + this.state.marginTop
            ]
          })
        }
      ]
    };
    return (
      <Animated.View
        onLayout={this.handleLayout}
        style={[styles.drawer, viewStyle]}>
        <View style={styles.headerView}>
          <FlatList
            ref={this.handleHeaderListRef}
            data={this.props.data}
            style={styles.header}
            renderItem={this.renderHeaderItem}
            extraData={this.state}
            horizontal={true}
            pagingEnabled={true}
            keyExtractor={keyExtractor}
            showsHorizontalScrollIndicator={false}
            onScroll={this.handleHeaderScroll}
          />
          {header}
        </View>
        <View style={[styles.detailsView]}>
          {this.props.selectedItem.type === "stop" ? (
            <Arrivals selectedStop={this.props.selectedItem} />
          ) : null}
        </View>
      </Animated.View>
    );
  }

  selectItem(idx) {
    this.props.onSelectItemIndex(idx);
  }
}

function mapStateToProps(state) {
  return {
    selectedItem: getSelectedItem(state),
    selectedItemIndex: state.selectedItemIndex
  };
}

function mapDispatchToProps(dispatch, ownProps) {
  return {
    onSelectItemIndex: itemIndex => {
      dispatch(selectItemIndex(itemIndex));
      if (ownProps.onResize) {
        ownProps.onResize(itemIndex);
      }
    },
    onResize: size => {
      dispatch(setMapViewInset(size));
      if (ownProps.onResize) {
        ownProps.onResize();
      }
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(SelectedItemsView);

function StopItem(props) {
  let {width, onPress, stop} = props;
  return (
    <View
      style={{
        flex: 1,
        flexDirection: "row",
        width
      }}>
      <TouchableOpacity onPress={onPress}>
        <View style={styles.itemImageContainer}>
          <Image style={styles.itemImage} source={stopImage} />
        </View>
      </TouchableOpacity>
      <TouchableOpacity onPress={onPress}>
        <View key={stop.name} style={[styles.itemDescription]}>
          <View>
            <Text style={{fontSize: 18}}>{stop.name}</Text>
          </View>
          <View>
            <Text
              style={{
                color: "#999999",
                fontSize: 14,
                justifyContent: "center"
              }}>
              {stop.desc}
            </Text>
          </View>
        </View>
      </TouchableOpacity>
    </View>
  );
}

function VehicleItem(props) {
  let {width, vehicle, onPress} = props;
  return (
    <View
      style={{
        flex: 1,
        flexDirection: "row",
        width
      }}>
      <TouchableOpacity onPress={onPress}>
        <View style={styles.itemImageContainer}>
          <Image
            style={styles.itemImage}
            source={VEHICLE_IMAGE[vehicle.route_type]}
          />
          <Text>{vehicle.vehicle.id}</Text>
        </View>
      </TouchableOpacity>
      <TouchableOpacity onPress={onPress}>
        <View
          key={vehicle.vehicle.id}
          style={[styles.itemDescription, {flex: 1, flexDirection: "row"}]}>
          <View
            style={{
              padding: 10
            }}>
            <Text style={{fontSize: 36}}>{vehicle.trip.route_id}</Text>
          </View>

          <View>
            <Text style={{fontSize: 20}}>{vehicle.vehicle.label}</Text>
            <Text style={{fontSize: 10}}>
              Updated: {Moment.unix(vehicle.timestamp).fromNow()}
            </Text>
          </View>
        </View>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  drawer: {
    flex: 1,
    position: "absolute",
    left: 0,
    bottom: 0,
    width: "100%",
    backgroundColor: "#FFFFFF",
    transform: [{perspective: 1000}]
  },
  header: {
    borderColor: "rgba(0, 0, 0, 0.1)",
    borderTopWidth: 1,
    bottom: 0,
    maxHeight: 100,
    height: 100,
    padding: 1
  },
  headerView: {
    borderBottomWidth: 1,
    borderColor: "#CCCCCC"
  },
  detailsView: {
    paddingTop: 10,
    paddingBottom: 10,
    paddingLeft: 5,
    paddingRight: 5,
    position: "relative",
    top: -8,
    flex: 1
  },
  headerPagination: {
    flexDirection: "row",
    position: "relative",
    maxHeight: 25,
    backgroundColor: "#44FFFF",
    marginLeft: 10,
    marginRight: 10
  },
  inactive: {
    alignItems: "center",
    backgroundColor: "#FFFFFF",
    justifyContent: "center",
    flex: 1
  },
  active: {
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#DDDDDD",
    flex: 1
  },
  itemDescription: {
    alignItems: "center",
    height: 100,
    paddingLeft: 10,
    paddingRight: 10,
    minHeight: 100,
    flex: 0
  },
  itemImageContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    width: 60,
    maxWidth: 60,
    height: 100,
    backgroundColor: "#EFEFEF"
  },
  itemImage: {
    marginTop: 0,
    padding: 0,
    width: 40,
    resizeMode: "contain",
    height: 40,
    alignItems: "center"
  }
});
