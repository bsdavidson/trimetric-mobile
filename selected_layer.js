import React, {Component} from "react";
import {StyleSheet, View} from "react-native";
import {connect} from "react-redux";

import Mapbox from "@mapbox/react-native-mapbox-gl";

import {MIN_LABEL_LAYER_ID} from "./constants";
import {getSelectedItem} from "./selectors";

export function SelectedLayer(props) {
  const {selectedItem} = props;
  if (!selectedItem || selectedItem.type === "vehicle") {
    return null;
  }
  return (
    <View>
      <Mapbox.ShapeSource id="pressed_points_source" shape={selectedItem.item}>
        <Mapbox.CircleLayer
          id="pressed_points_layer"
          style={mapStyles.pressedPointsLayer}
          belowLayerID={MIN_LABEL_LAYER_ID}
        />
      </Mapbox.ShapeSource>
    </View>
  );
}

function mapStateToProps(state) {
  return {
    selectedItem: getSelectedItem(state)
  };
}

export default connect(mapStateToProps)(SelectedLayer);

const mapStyles = Mapbox.StyleSheet.create({
  pressedPointsLayer: {
    circleColor: "rgba(0,0,0,0)",
    circleStrokeWidth: 2,
    circleStrokeOpacity: 0.5,
    circleStrokeColor: "#00FFFF",
    circleRadius: 14
  },
  pressLineLayer: {
    lineColor: "#00FF00",
    lineOpacity: 1,
    lineWidth: 3
  }
});
