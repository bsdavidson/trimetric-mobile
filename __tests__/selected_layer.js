import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import {SelectedLayer} from "../selected_layer";
import {selectedItem} from "../__fixtures__/selected_items_view";

jest.mock("@mapbox/react-native-mapbox-gl");

it("renders selectedLayer", () => {
  const slNull = renderer.create(<SelectedLayer />).toJSON();
  expect(slNull).toMatchSnapshot();

  const sl = renderer
    .create(<SelectedLayer selectedItem={selectedItem} />)
    .toJSON();
  expect(sl).toMatchSnapshot();
});
