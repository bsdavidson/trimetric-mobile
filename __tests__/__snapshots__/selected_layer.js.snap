// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders selectedLayer 1`] = `null`;

exports[`renders selectedLayer 2`] = `
<View>
  <ShapeSource
    id="pressed_points_source"
    shape={
      Object {
        "geometry": Object {
          "coordinates": Array [
            -122.68415115773678,
            45.50685913695389,
          ],
          "type": "Point",
        },
        "properties": Object {
          "stop_id": "7749",
          "type": "stop",
        },
        "type": "Feature",
      }
    }
  >
    <CircleLayer
      belowLayerID="place-city-sm"
      id="pressed_points_layer"
      style={
        Object {
          "circleColor": "rgba(0,0,0,0)",
          "circleRadius": 14,
          "circleStrokeColor": "#00FFFF",
          "circleStrokeOpacity": 0.5,
          "circleStrokeWidth": 2,
        }
      }
    />
  </ShapeSource>
</View>
`;
